tool/tsh/kube.go

```diff
@@ -0,0 +1,269 @@
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/client-go/pkg/apis/clientauthentication"
+	clientauthv1beta1 "k8s.io/client-go/pkg/apis/clientauthentication/v1beta1"
+)
+
+type kubeCommands struct {
+	credentials *kubeCredentialsCommand
+	clusters    *kubeClustersCommand
+	login       *kubeLoginCommand
+}
+
+func newKubeCommand(app *kingpin.Application) kubeCommands {
+	kube := app.Command("kube", "Manage available kubernetes clusters").Hidden()
+	cmds := kubeCommands{
+		credentials: newKubeCredentialsCommand(kube),
+		clusters:    newKubeClustersCommand(kube),
+		login:       newKubeLoginCommand(kube),
+	}
+	return cmds
+}
+
+type kubeCredentialsCommand struct {
+	*kingpin.CmdClause
+	kubeCluster     string
+	teleportCluster string
+}
+
+func newKubeCredentialsCommand(parent *kingpin.CmdClause) *kubeCredentialsCommand {
+	c := &kubeCredentialsCommand{
+		// This command is always hidden. It's called from the kubeconfig that
+		// tsh generates and never by users directly.
+		CmdClause: parent.Command("credentials", "Get credentials for kubectl access").Hidden(),
+	}
+	c.Flag("teleport-cluster", "Name of the teleport cluster to get credentials for.").Required().StringVar(&c.teleportCluster)
+	c.Flag("kube-cluster", "Name of the kubernetes cluster to get credentials for.").Required().StringVar(&c.kubeCluster)
+	return c
+}
+
+func (c *kubeCredentialsCommand) run(cf *CLIConf) error {
+	tc, err := makeClient(cf, true)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// Try loading existing keys.
+	k, err := tc.LocalAgent().GetKey(c.teleportCluster)
+	if err != nil && !trace.IsNotFound(err) {
+		return trace.Wrap(err)
+	}
+	// Loaded existing credentials and have a cert for this cluster? Return it
+	// right away.
+	if err == nil {
+		crt, err := k.KubeTLSCertificate(c.kubeCluster)
+		if err != nil && !trace.IsNotFound(err) {
+			return trace.Wrap(err)
+		}
+		if crt != nil && time.Until(crt.NotAfter) > time.Minute {
+			log.Debugf("Re-using existing TLS cert for kubernetes cluster %q", c.kubeCluster)
+			return c.writeResponse(k, c.kubeCluster)
+		}
+		// Otherwise, cert for this k8s cluster is missing or expired. Request
+		// a new one.
+	}
+
+	log.Debugf("Requesting TLS cert for kubernetes cluster %q", c.kubeCluster)
+	err = client.RetryWithRelogin(cf.Context, tc, func() error {
+		return tc.ReissueUserCerts(cf.Context, client.ReissueParams{
+			RouteToCluster:    c.teleportCluster,
+			KubernetesCluster: c.kubeCluster,
+		})
+	})
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// ReissueUserCerts should cache the new cert on disk, re-read them.
+	k, err = tc.LocalAgent().GetKey(c.teleportCluster)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	return c.writeResponse(k, c.kubeCluster)
+}
+
+func (c *kubeCredentialsCommand) writeResponse(key *client.Key, kubeClusterName string) error {
+	crt, err := key.KubeTLSCertificate(kubeClusterName)
+	if err != nil {
+		return trace.Wrap(err)
```

Comment:
Is there any specific error output format for the credential's plugin response or kubectl solely relies on exit code, do you know?